### ISSUE TITLE
Rework ZeebeDB API

### ISSUE THREAD

**Issue by ChrisKujawa** [camunda/camunda#12595]
### Description:

It is more (?) than three years ago when we created the ZeebeDB API, and I feel it needs a new coat of whitewash.

The ZeebeDB API was designed based on our usage and was the first attempt at designing the API. I think we have learned something new now. For example, the current key-value pairs and in general the ColumnFamily design causes a lot of boilerplate code. Furthermore, it contains side effects that might sometimes not be clear. The key value objects which are used on creation for the column family are changing when retrieving data etc.

I think it is possible to simplify the API in a way that we can reduce the side effects and boilerplate code plus improve the usability. If we can reduce a lot of the boilerplate in the state classes I think we can increase the maintability.

### Example POC

I did a small POC to redesign the API. This is just one of many potential solutions, it doesn't mean we need to do it like that.

Most of the time is about moving the complexity in this case we would move the complexity into the ZeebeDB impl. We could define TypeEncoderDecoder, which is registered internally and based on the type of the key and values, the right TypeEncoderDecoder are chosen. It would allow to use the key-values directly on insert, get etc.

See the branch https://github.com/camunda/zeebe/commits/zell-refactor-zbdb-api

For Strings, CompositeKeys, etc it should be easily doable to create such TypeEncoderDecoder. Either we provide some for BufferWriter/Reader as well or the user of ZeebeDB API might bring their own TypeEncoderDecoder in order to support also our other value objects.

#### Example for String:

```java

  // BEFORE
  @Test
  public void shouldUseOldAPI() {
    // given
    final DbString keyString = new DbString();
    final DbString valueString = new DbString();
    final var stringStringCF =
        zeebeDb.createColumnFamily(
            DefaultColumnFamily.DEFAULT, zeebeDb.createContext(), keyString, valueString);

    // when
    keyString.wrapString("hallo");
    valueString.wrapString("du");
    stringStringCF.insert(keyString, valueString);
    valueString.wrapString("test");

    // then
    final var value = stringStringCF.get(keyString);
    assertThat(value.toString()).isEqualTo("du");
    assertThat(valueString.toString()).isEqualTo("du");
  }

  // AFTER
  @Test
  public void shouldUseNewAPI() {
    // given
    final var stringStringNewColumnFamily =
        zeebeDb.<String, String>newCreateColumnFamily(
            DefaultColumnFamily.DEFAULT, zeebeDb.createContext());

    // when
    stringStringNewColumnFamily.insert("hallo", "du");

    // then
    final String value = stringStringNewColumnFamily.get("hallo");
    assertThat(value).isEqualTo("du");
  }
```

#### Example composite keys

Most interesting are composite keys, which we have a lot in our code base. But they use mostly also just basic types like long or string.

```java
  // BEFORE
  @test
  public void shouldUseOldAPICompositeKeys() {
    // given
    final DbString firstKey = new DbString();
    final DbString secondKey = new DbString();
    final DbString stringValue = new DbString();
    final var compositeKey = new DbCompositeKey<>(firstKey, secondKey);
    final var stringStringCF =
        zeebeDb.createColumnFamily(
            DefaultColumnFamily.DEFAULT, zeebeDb.createContext(), compositeKey, stringValue);

    // when
    firstKey.wrapString("hallo");
    secondKey.wrapString("du");
    stringValue.wrapString("!!!");
    stringStringCF.insert(compositeKey, stringValue);

    // then
    final var value = stringStringCF.get(compositeKey);
    assertThat(value.toString()).isEqualTo("!!!");
  }

  // AFTER
  @test
  public void shouldUseNewAPICompositeKeys() {
    // given
    final var stringStringCF =
        zeebeDb.<StringStringComposite, String>newCreateColumnFamily(
            DefaultColumnFamily.DEFAULT, zeebeDb.createContext());

    // when
    stringStringCF.insert(StringStringComposite.of("hallo", "du"), "!!!");

    // then
    final var value = stringStringCF.get(StringStringComposite.of("hallo", "du"));
    assertThat(value).isEqualTo("!!!");
  }
```

@camunda/zeebe-process-automation @camunda/zeebe-distributed-platform I put it on both boards, since the responsibility still needs to be defined/discussed again, see #12203. Furthermore, ZPA is the consumer of the API so they are the most affected by this, which needs to give input or drive this.

**Comment by npepinpe** [comment]
:rocket: Moving away from self-serializing types to picking the right serializers at construction time and just passing in the values is the best :+1: Not only does it reduce boilerplate, but it will also prevent errors due to reuse of the type instances (e.g. forget to clear the previous `DbString` value and it gets reuse indirectly or something).

**Comment by npepinpe** [comment]
I also have a pet peeve, which is that we use the name `ColumnFamily` for things that aren't column families :smile: I would love if we changed the name to something non RocksDB specific, but that's very minor, just a minor annoyance from my side :sweat_smile:

**Comment by npepinpe** [comment]
@Zelldon - consensus is to put it in ready and within the scope of the large state topic (to simplifying caching), but I'll let you decide as DRI to put it in the ready column/treat it as the next immediate issue for this topic

**Comment by Zelldon** [comment]
I don't think that this is in the scope of #12033. I think it makes sense to tackle after #12203 for sure and then focused.

**Comment by marcosgvieira** [comment]
### 🧭 Triage Assessment

| Field | Value |
|------|-------|
| **Work type** | TASK_TOIL_NON_FEATURE |
| **Impact** | 🔥 High |
| **When** | ⚠️ Soon |
| **Confidence** | 0.91 |

---

### 📝 Rationale

Internal ZeebeDB API refactoring/technical debt to reduce boilerplate, side effects, and maintenance burden; it is not incorrect released behavior or a new product capability. The API is consumed by ZPA and spans storage/API boundaries, so impact is high, while delay creates avoidable engineering friction. 

---

🤖 _This issue was automatically triaged by an AI-based system as part of the Core Features triage workflow._

✋ **Want a human review?**
If you believe this assessment is incorrect or needs adjustment, feel free to update the details or request a manual triage by adding the **`triage:manual`** label.
